File: recipe-java-runtime/src/main/java/ca/derekcormier/recipe/BackendOven.java
package ca.derekcormier.recipe;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.jsontype.NamedType;
import com.fasterxml.jackson.databind.jsontype.SubtypeResolver;
import com.fasterxml.jackson.databind.jsontype.impl.StdSubtypeResolver;

import java.util.HashMap;
import java.util.Map;

public class BackendOven extends AbstractOven {
    private final Map<String,BaseIngredientHook> hooks = new HashMap<>();
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final SubtypeResolver subtypeResolver = new StdSubtypeResolver();

    public BackendOven() {
        objectMapper.setSubtypeResolver(subtypeResolver);
        subtypeResolver.registerSubtypes(new NamedType(Recipe.class, "Recipe"));
    }

    public String bake(String json) {
        try {
            Payload payload = objectMapper.readValue(json, Payload.class);
            Cake cake = createCake(payload.getCake());

            bakeIngredient(payload.getRecipe(), cake);

            Cake cakeToSerialize = new Cake(cake);
            return objectMapper.writeValueAsString(cakeToSerialize);
        }
        catch (Exception e) {
            throw new RuntimeException("payload serialization error", e);
        }
    }

    public void registerHook(BaseIngredientHook hook) {
        hooks.put(hook.getIngredientName(), hook);
        subtypeResolver.registerSubtypes(new NamedType(hook.getDataClass(), hook.getIngredientName()));
    }

    private void bakeIngredient(Ingredient ingredient, Cake cake) {
        if (ingredient instanceof Recipe) {
            Recipe recipe = (Recipe)ingredient;

            Runnable bakeRecipeIngredients = () -> {
                for (Ingredient i: recipe.getIngredients()) {
                    bakeIngredient(i, cake);
                }
            };

            if (recipe.getContext() != null) {
                cake.inNamespace(recipe.getContext(), bakeRecipeIngredients);
            }
            else {
                bakeRecipeIngredients.run();
            }
        }
        else {
            hooks.get(ingredient.getIngredientType()).bake(ingredient, cake);
        }
    }
}

File: recipe-ts-runtime/src/BackendOven.ts
import { AbstractOven } from "./AbstractOven";
import { Payload } from "./Payload";
import { Cake } from "./Cake";
import { BaseIngredientHook } from "./BaseIngredientHook";
import { Ingredient } from "./Ingredient";
import { Recipe } from "./Recipe";

export class BackendOven extends AbstractOven {
    private readonly hooks: {[name: string]: BaseIngredientHook<any>} = {};
    private readonly ingredientTypes: {[name: string]: Function} = {};

    public bake(json: string): string {
        const payload = Payload.fromJSON(JSON.parse(json), this.ingredientTypes);
        const cake = this.createCake(payload.getCake());

        this.bakeIngredient(payload.getRecipe(), cake);

        const cakeToSerialize = new Cake(cake);
        return JSON.stringify(cakeToSerialize);
    }

    public registerHook<T>(hook: BaseIngredientHook<T>) {
        this.hooks[hook.getIngredientName()] = hook;
        this.ingredientTypes[hook.getIngredientName()] = hook.getSnapshotClass();
    }

    private bakeIngredient(ingredient: Ingredient, cake: Cake) {
        if (ingredient instanceof Recipe) {
            const recipe = ingredient as Recipe;

            const bakeRecipeIngredients = () => {
                for (let i of recipe.getIngredients()) {
                    this.bakeIngredient(i, cake);
                }
            };

            if (recipe.getContext() !== null) {
                cake.inNamespace(recipe.getContext()!, bakeRecipeIngredients);
            }
            else {
                bakeRecipeIngredients();
            }
        }
        else {
            const ingredientData = ingredient.toJSON()[ingredient.getIngredientType()];
            this.hooks[ingredient.getIngredientType()].bake(ingredientData, cake);
        }
    }
}